iroha_data_model/src/events.rs

```diff
@@ -244,101 +265,96 @@ pub mod pipeline {
         /// Error which happened during execution
         pub reason: String,
     }
+    impl std::fmt::Display for InstructionExecutionFail {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            use Instruction::*;
+            let type_ = match self.instruction {
+                Burn(_) => "burn",
+                Fail(_) => "fail",
+                If(_) => "if",
+                Mint(_) => "mint",
+                Pair(_) => "pair",
+                Register(_) => "register",
+                Sequence(_) => "sequence",
+                Transfer(_) => "transfer",
+                Unregister(_) => "unregister",
+            };
+            write!(
+                f,
+                "Failed execute instruction of type {}: {}",
+                type_, self.reason
+            )
+        }
+    }
+    impl std::error::Error for InstructionExecutionFail {}
 
     /// Transaction was reject because of low authority
     #[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Decode, Encode)]
     pub struct NotPermittedFail {
         /// Reason of failure
         pub reason: String,
     }
+    impl std::fmt::Display for NotPermittedFail {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            write!(f, "Action not permitted: {}", self.reason)
+        }
+    }
+    impl std::error::Error for NotPermittedFail {}
 
     /// The reason for rejecting transaction which happened because of new blocks.
     #[derive(
-        Debug, Clone, Copy, Eq, PartialEq, Serialize, Deserialize, Decode, Encode, FromVariant,
+        Debug,
+        Clone,
+        Copy,
+        Eq,
+        PartialEq,
+        Serialize,
+        Deserialize,
+        Decode,
+        Encode,
+        FromVariant,
+        Error,
     )]
     pub enum BlockRejectionReason {
         /// Block was rejected during consensus.
         //TODO: store rejection reasons for blocks?
+        #[error("Block was rejected during consensus.")]
         ConsensusBlockRejection,
     }
 
     /// The reason for rejecting transaction which happened because of transaction.
-    #[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Decode, Encode, FromVariant)]
+    #[derive(
+        Debug, Clone, Eq, PartialEq, Serialize, Deserialize, Decode, Encode, FromVariant, Error,
+    )]
     pub enum TransactionRejectionReason {
         /// Failed due to low authority.
-        NotPermitted(NotPermittedFail),
+        #[error("Transaction rejected due to low authority")]
+        NotPermitted(#[from] NotPermittedFail),
```

Comment:
Do we need `#[from]` if we have `FromVariant`?

Reply:
Ambiguous name, but it actually says to macro to make it `Error::source` for this variant.

Reply:
What do you think, should we rename it to `source`?